### ISSUE TITLE
Integer conversion in SYCL floating-point math functions

### ISSUE THREAD

**Issue by steffenlarsen** [KhronosGroup/SYCL-Docs#417]
The [C++ draft for cmath](https://eel.is/c++draft/c.math#cmath.syn-3) states that:

> For each function with at least one parameter of type floating-point-type other than abs, the implementation also provides additional overloads sufficient to ensure that, if every argument corresponding to a floating-point-type parameter has arithmetic type, then every such argument is effectively cast to the floating-point type with the greatest floating-point conversion rank and greatest floating-point conversion subrank among the types of all such arguments, where arguments of integer type are considered to have the same floating-point conversion rank as double[.](https://eel.is/c++draft/c.math#cmath.syn-3.sentence-1) If no such floating-point type with the greatest rank and subrank exists, then overload resolution does not result in a usable candidate ([[over.match.general]](https://eel.is/c++draft/over.match.general)) from the overloads provided by the implementation[.](https://eel.is/c++draft/c.math#cmath.syn-3.sentence-2)

However, SYCL 2020 makes no such guarantees about overloads for its math functions, effectively meaning calling floating-point math functions (e.g. `sycl::cos`) with an integer may cause ambiguity errors as either the `float` or `double` overloads are valid for it, while the corresponding cmath math functions (e.g. `std::cos`) would follow the aforementioned rules and pick the appropriate overload accordingly.

Should SYCL adopt similar rules to make porting from cmath to the equivalent SYCL math functions simpler?

**Comment by gmlueck** [comment]
I think this issue only exists for the SYCL math functions that take all scalar parameters.  For example:

```
// In <sycl/sycl.hpp>
namespace sycl {

float cos(float);
double cos(double);
float fmax(float, float);
double fmax(double, double);

}

int val = 2;
sycl::cos(1);        // Error (A)
sycl::fmax(val, 1);  // Error (B)
```

Both (A) and (B) will cause an error due to ambiguous overload resolution because the integer values (`1` and `val`) could be implicitly converted to either `float` or `double`.  There is no similar ambiguity for the functions that take an `marray` or `vec` parameter:

```
// In <sycl/sycl.hpp>
namespace sycl {

template<size_t N>
marray<float,N> fmax(marray<float,N>, float);
template<size_t N>
marray<double,N> fmax(marray<double,N>, double);

}

sycl::marray<float, 2> maf{1.0f, 2.0f};
sycl::marray<int, 2> mai{1, 2};
sycl::fmax(maf, 1);  // OK (A)
sycl::fmax(mai, 1);  // Error (B)
```

In this case, (A) is not ambiguous because the `maf` parameter selects the `float` overload, so the integer value `1` is converted to `float`.  Case (B) is not ambiguous, but it's still an error because there is no implicit conversion from `marray<int,N>` to either `marray<float,N>` or `marray<double,N>`.

Therefore, it seems like the concern raised in this issue only exists for the math functions that take all scalar arguments.

Why do we want these functions in SYCL at all, though?  Wouldn't it be better to rely on the standard C++ functions like `std::cos` and `std::fmax` instead?  Therefore, I think the resolution of this issue should be to deprecate (and eventually remove) the SYCL math functions that take all scalar arguments.

**Comment by Pennycook** [comment]
> Why do we want these functions in SYCL at all, though? Wouldn't it be better to rely on the standard C++ functions like `std::cos` and `std::fmax` instead? Therefore, I think the resolution of this issue should be to deprecate (and eventually remove) the SYCL math functions that take all scalar arguments.

I don't think we can guarantee that the `std::` functions work in device code for arbitrary C++ standard library implementations.  Wouldn't requiring certain `std::` functions to work require some changes to conformance? (e.g. "this compiler is conformant with this/these versions of a C++ library and headers")

As a hypothetical example: the host implementation of `std::cos` might be implemented via something like a call to a `__builtin_cos` intrinsic, if those headers are designed to work only with a subset of host compilers.  A device compiler would need to understand and implement `__builtin_cos` for compilation to succeed.

I'm not opposed to this as a long-term direction, but it doesn't seem like a quick fix.  I think in the short-term we should say that the `sycl::` math functions are guaranteed to have behavior that is consistent with C++.  If we later decide to deprecate the `sycl::` versions in favor of the `std::` versions, the migration would be much smoother.

**Comment by gmlueck** [comment]
We have been talking about providing a guarantee in SYCL-Next that allows some subset of C++ functions to be used in device code.  I assumed that these math functions would be non-controversial candidates for that list.  (Maybe not true?)  DPC++ does already guarantee that most standard math functions work in device code.

**Comment by Pennycook** [comment]
> We have been talking about providing a guarantee in SYCL-Next that allows some subset of C++ functions to be used in device code. I assumed that these math functions would be non-controversial candidates for that list. (Maybe not true?) DPC++ does already guarantee that most standard math functions work in device code.

Honestly, I'm not sure if they would be controversial. All I feel comfortable saying is that I wouldn't be surprised if math functions were more likely to include compiler-specific built-ins than other C++ standard library functions (say, `std::swap`).

The DPC++ documentation might be outdated, but according to [the extension](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/supported/C-CXX-StandardLibrary.rst) the standard math functions which are supported only work if you're using GNU libc or Microsoft's C library.

**Comment by gmlueck** [comment]
We should also decide whether this is a "bug fix" to SYCL 2020 vs. a SYCL-Next thing.  I've been thinking it is a SYCL-Next thing.

**Comment by TApplencourt** [comment]
I have little experience porting math `std::` to SYCL, we did that for std::complex. 

I don't really see how `library-only` people will be able to use `std::lib`. Indeed, they have no control over it. So even if one intrinsics exist they will be not able to compile them. 

In practice, it may work for most of the libc (they try to `ifdef` all the intrinsic, and sometimes even support include in cuda code) but, sadly, I'm not sure we can make it mandatory for implementers to support std function.

I prefer the John approach:
> think in the short-term we should say that the sycl:: math functions are guaranteed to have behavior that is consistent with C++.
